packages/dds/sequence/src/test/rebasing.spec.ts

```diff
@@ -0,0 +1,117 @@
+ * Licensed under the MIT License.
+ */
+
+import { strict as assert } from "assert";
+import {
+	MockFluidDataStoreRuntime,
+	MockContainerRuntimeFactory,
+	MockContainerRuntime,
+	MockStorage,
+} from "@fluidframework/test-runtime-utils";
+import { IMergeTreeInsertMsg } from "@fluidframework/merge-tree";
+import { FlushMode } from "@fluidframework/runtime-definitions";
+import { SharedString } from "../sharedString";
+import { SharedStringFactory } from "../sequenceFactory";
+
+[
+	{
+		options: {
+			flushMode: FlushMode.Immediate,
+		},
+		name: "FlushMode immediate",
+	},
+	{
+		options: {
+			flushMode: FlushMode.TurnBased,
+			enableGroupedBatching: true,
+		},
+		name: "FlushMode TurnBased with grouped batching",
+	},
+].forEach((testConfig) => {
+	describe(`Rebasing - ${testConfig.name}`, () => {
+		let containerRuntimeFactory: MockContainerRuntimeFactory;
+		let containerRuntime1: MockContainerRuntime;
+		let containerRuntime2: MockContainerRuntime;
+		let sharedString1: SharedString;
+		let sharedString2: SharedString;
+
+		const createSharedString = async (
+			id: string,
+			factory: MockContainerRuntimeFactory,
+		): Promise<[SharedString, MockContainerRuntime]> => {
+			const dataStoreRuntime = new MockFluidDataStoreRuntime();
+			dataStoreRuntime.local = false;
+			const containerRuntime = factory.createContainerRuntime(dataStoreRuntime);
+			const services = {
+				deltaConnection: containerRuntime.createDeltaConnection(),
+				objectStorage: new MockStorage(),
+			};
+			const sharedString = new SharedString(
+				dataStoreRuntime,
+				id,
+				SharedStringFactory.Attributes,
+			);
+			sharedString.initializeLocal();
+			sharedString.connect(services);
+			return [sharedString, containerRuntime];
+		};
+
+		beforeEach(async () => {
+			containerRuntimeFactory = new MockContainerRuntimeFactory(testConfig.options);
+			[sharedString1, containerRuntime1] = await createSharedString(
+				"shared-string-1",
+				containerRuntimeFactory,
+			);
+			[sharedString2, containerRuntime2] = await createSharedString(
+				"shared-string-2",
+				containerRuntimeFactory,
+			);
+		});
+
+		it("Rebasing ops maintains eventual consistency", async () => {
+			assert(containerRuntime1.rebase !== undefined, "Unsupported test-utils version");
+			assert(containerRuntime1.flush !== undefined, "Unsupported test-utils version");
+			assert(containerRuntime2.rebase !== undefined, "Unsupported test-utils version");
+			assert(containerRuntime2.flush !== undefined, "Unsupported test-utils version");
+
+			sharedString1.insertText(0, "ad");
+			sharedString1.insertText(1, "c");
+			containerRuntime1.flush();
+			containerRuntime2.flush();
+			containerRuntimeFactory.processAllMessages();
+
+			sharedString2.on("sequenceDelta", (sequenceDeltaEvent) => {
+				if ((sequenceDeltaEvent.opArgs.op as IMergeTreeInsertMsg).seg === "b") {
+					sharedString2.insertText(3, "u");
+					containerRuntime2.rebase?.();
```

Comment:
given the context (only one op is submitted), I don't think this matters much, but would it be more correct to call `rebase` after `processOneMessage` in the outer loop?

Asking as I'm imagining that the prod codepath calls rebase immediately after finishing processing a batch of messages, not after each op that requires a rebase. but perhaps that's not how it works/will work.

Reply:
Good point. Rebasing should actually happen before the flush.